#10 ClassCastException when in RouteResolverException is checked

mbyrne00 opened the pull request:
* Pulled `ViewResolverNotFoundException` into its own `catch` since there was no need to check instance to just re-throw
 * Pass `Throwable` when attempting to resolve an exception view - still throw original exception if none can be found

atomicleopard commented:
Hi Matt, looking at this change I think that there are some unintended changes included that might cause issues. The previous behaviour permitted the ViewResolverNotFoundException to be the unwrapped RuntimeException which is no longer true.

Exception handling has been a bit hairy in the past because of how servlet containers handle them, so it wouldn't surprise me if that logic was hard won. It can be a bit challenging to test because the behaviour changes based on the container. None of the tests leap out at me as covering a specific scenario though.

Incidentally, the call in the catch block to `resp.finaliseHeaders();` could also be problematic, if the resp was successfully finalised by the viewresolver (which is part of the contract).

Probably two options:

* keep the behaviour as it was rather than separating out the separate catch block
* keep your change but test four basic scenarios in a test app - a viewresolverexception that causes a servlet resp.sendError() that delegates to a web.xml configured error page and one that renders a StringView, and another type of runtime exception that causes the same. If those all work as expected, it should be good to go.

mbyrne00 commented:
Thanks Nick. I'll take a look in more detail tomorrow and try to test as
you suggested. It sounds like I could add one or two more unit tests to
cover some things you mentioned

On Wednesday, 21 September 2016, Atomic Leopard <notifications@github.com>
wrote:

> Hi Matt, looking at this change I think that there are some unintended
> changes included that might cause issues. The previous behaviour permitted
> the ViewResolverNotFoundException to be the unwrapped RuntimeException
> which is no longer true.
>
> Exception handling has been a bit hairy in the past because of how servlet
> containers handle them, so it wouldn't surprise me if that logic was hard
> won. It can be a bit challenging to test because the behaviour changes
> based on the container. None of the tests leap out at me as covering a
> specific scenario though.
>
> Incidentally, the call in the catch block to resp.finaliseHeaders();
> could also be problematic, if the resp was successfully finalised by the
> viewresolver (which is part of the contract).
>
> Probably two options:
>
>    - keep the behaviour as it was rather than separating out the separate
>    catch block
>    - keep your change but test four basic scenarios in a test app - a
>    viewresolverexception that causes a servlet resp.sendError() that delegates
>    to a web.xml configured error page and one that renders a StringView, and
>    another type of runtime exception that causes the same. If those all work
>    as expected, it should be good to go.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/3wks/thundr/pull/11#issuecomment-248590076>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFbKrpFWjKAczqF34VVGJxDpg_3-pHWBks5qsRvFgaJpZM4KCLvM>
> .
>


-- 
Sent from my phone

mbyrne00 commented:
Hey Nick,

The separate catch block seemed like it would behave the same but I missed the subtlety there. I've added a test to assert this and put it back how you had that part. Just to make sure things were exactly like before I've also tried to throw the `RouteResolverException` cause if there is no matching error view and the cause is `RuntimeException` - that way it behaves like before. If the cause is checked I just throw the original exception. This way everything should behave as it did before - the only new behaviour is for a checked cause. Tests should cover this too - verified that only my checked exception tests fail if I reverted my code changes, so that confirms existing behaviour.

mbyrne00 commented:
Manual testing done with built snapshot and I can confirm that exception view resolving via a customer Filter we wrote in our app works before and after this fix. I was not able to get `web.ml` `error-page` to work *with or without* my change. I tried using the approach documented here http://3wks.github.io/thundr/2.0/views/httpstatusexception.html (using `StatusCode` as first param instead of `int`).

mbyrne00 commented:
Reviewed with @shanebell and merging into 3.0 branch.

atomicleopard commented:
Re: error pages - what container were you using? And were you using thundr-servlet-25 or thundr-servlet-31? It's pretty fundamental that error pages work.

mbyrne00 commented:
Was gae, so would have been servlet-25 I guess.

On 12 October 2016 at 17:20, Atomic Leopard <notifications@github.com>
wrote:

> Re: error pages - what container were you using? And were you using
> thundr-servlet-25 or thundr-servlet-31? It's pretty fundamental that error
> pages work.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/3wks/thundr/pull/11#issuecomment-253129195>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFbKrsutRhQ6b4BMK8CumqlmWpu8Akldks5qzHwpgaJpZM4KCLvM>
> .
>
